<a href="https://colab.research.google.com/github/paulinawdowiak1/ML-Bootcamp/blob/main/supervised/ML_Bootcamp_linear_regression_normal_equation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np

In [2]:
!pip install scikit-learn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


### <a name='1'></a> Przygotowanie danych

In [3]:
X1 = np.array([1, 2, 3, 4, 5, 6])
Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
m = len(X1)

print(f'Lata pracy: {X1}')
print(f'Wynagrodzenie: {Y}')
print(f'Liczba próbek: {m}')

Lata pracy: [1 2 3 4 5 6]
Wynagrodzenie: [3000 3250 3500 3750 4000 4250]
Liczba próbek: 6


In [4]:
X1 = X1.reshape(m, 1)
print(X1)
print(X1.shape)

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]
(6, 1)


In [6]:
#generowanie wektora 1
bias = np.ones((m, 1))
print(bias)
print(bias.shape)

[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]
(6, 1)


In [7]:
X = np.append(bias, X1, axis=1)
print(X)
print(X.shape)

[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]
(6, 2)


### <a name='2'></a> Równanie normalne

Regresja liniowa w $R^2$:
$$Y = w_0 + w_1X_1$$
$$Y = W^TX$$
gdzie: $$W= \left[\begin{matrix}w_0\\w_1\end{matrix}\right]$$ $$ X= \left[\begin{matrix}1\\X_1\end{matrix}\right] $$ stąd $$ W^T= \left[\begin{matrix}w_0&w_1\end{matrix}\right] $$    
$$Y = W^TX = \left[\begin{matrix}w_0&w_1\end{matrix}\right] \cdot  \left[\begin{matrix}1\\X_1\end{matrix}\right] = w_0 + w_1X_1$$

Równanie normalne - równanie pozwalające obliczyć minimum funkcji straty (o ile istnieje)
$$W = (X^TX)^{-1}(X^TY)$$

In [8]:
#liczenie iloczynu z nawiasu pierwszego W=(XTX)
#np.dot - funkcja mnożenia macierzy
np.dot(X.T, X)

array([[ 6., 21.],
       [21., 91.]])

In [11]:
np.linalg?
#np.linalg.inv - wyznacza macierz odwrotna

In [9]:
#liczenie W=(XTX)−1 / odwrócenie macierzy
L = np.linalg.inv(np.dot(X.T, X))
L

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

In [10]:
#liczenie (XTY)
P = np.dot(X.T, Y)
P

array([21750., 80500.])

In [12]:
#wyliczenie wektora W
np.dot(L,P)

array([2750.,  250.])

Końcowa postać modelu$$Y = 2750 + 250X_1$$

### <a name='3'></a> Regresja liniowa przy pomocy scikit-learn

In [13]:
from sklearn.linear_model import LinearRegression

#w metodzie .fit nie trzeba tworzyc onesów, można działać od razu na macierzy X1

regression = LinearRegression()
regression.fit(X1, Y)

print(regression.intercept_) #W0(współczynnik W zero) - miejsce przecięcia z osią y
print(regression.coef_[0]) #w1

2750.0
250.0
